lib/personService.js

```diff
@@ -0,0 +1,41 @@
+const { Person } = require('../models/schema');
+
+class PersonService {
+  /**
+   * Create a new person.
+   * @param {object} bot - instance of slack bot.
+   * @param {string} userSlackId - slack id of the new person.
+   * @returns {Promise} request of creating new person query.
+  */
+  createPerson (bot, userSlackId) {
+    const promise = (resolve, reject) => {
+      // Get Slack user info through Botkit.
+      bot.api.users.info({ user: userSlackId }, async (err, res) => {
+        if (err) {
+          return reject(err);
+        }
+        const personRecord = Person.query().insertAndFetch({ slack_id: userSlackId, username: res.user.name });
+        resolve(personRecord);
+      });
+    };
+
+    return new Promise(promise);
+  }
```

Comment:
This is pretty nightmareish. There is a mix of async/await, promises and callbacks. This is what is called [releasing the Zalgo](https://www.quora.com/What-is-Zalgo-in-the-context-of-Node-js)

I understand why. The bot uses callbacks which makes it hard to do anything other than creating a wrapping promise with a resolve/reject.

The `async` on line 13 seem unnecessary unless the `Person.query()` on line 17 is async, in which case it's missing the `await` (but it's being returned so it will resolve anyway).

See what I mean about nightmareish? It's really unclear what is expected here.

The last thing I wanted to mention with this approach is that while in your success case and your handled error cases you return a promise (that resolves or rejects) in all other cases it will `throw` which means the caller needs to expect a promise as a return type but also expect an exception. That will mean the caller will do stuff like this.


```javascript
try {
  personService.createPerson(bot, 'this-is-an-id')
    .then(() => console.log('yay'))
    .catch(err => console.error('An expected error occurred', err))
} catch(err) {
  console.error('An unexpected error occurred', err)
}
```

My suggestion is to make the bot be async. There are a couple ways you can do this easily.
  - [promiseify the bot api](https://nodejs.org/dist/latest-v8.x/docs/api/util.html)
  - write a wrapper around `botkit` that becomes the new `bot`, it can bury the callback in there

Super happy to talk more about this is anyone cares to.

Reply:
@tonytamps thanks for your very detailed review! Just had a look at Botkit source code and seems quite difficult to make a wrapper around `botkit`(everything are based on function, and lots of nested callback functions), but would try `promisify` 👍

Reply:
Oh I must meant a file like this:

```javascript
// newBot.js
const botkit = require('botkit');

export async getInfo (user) {
  return new Promise((resolve, reject) => {
    bot.api.users.info({ user }, (err, res) => {
        if (err) return reject(err);

        const personRecord = Person.query().insertAndFetch({ slack_id: userSlackId, username: res.user.name });
        resolve(personRecord);
      });
  })

```

Reply:
(as discussed)